### ISSUE TITLE
remove use of CDNs from default themes

### ISSUE THREAD

**Issue by fkooman** [mkdocs/mkdocs#2171]
mkdocs looks like the perfect tool to self-host documentation! I'm a bit confused as to why CDNs are used for fonts and CSS. I'd prefer not to throw my visitors under the Google or Cloudflare bus :) Would it be possible to change this *by default* and not have the fonts and CSS come from CDNs? This is already done for some fonts and CSS files...

References:

https://github.com/mkdocs/mkdocs/blob/273e3baa8e47fa5d121402d98757d3f5a5510383/mkdocs/themes/readthedocs/base.html#L21

https://github.com/mkdocs/mkdocs/blob/273e3baa8e47fa5d121402d98757d3f5a5510383/mkdocs/themes/readthedocs/base.html#L26

https://github.com/mkdocs/mkdocs/blob/273e3baa8e47fa5d121402d98757d3f5a5510383/mkdocs/themes/readthedocs/base.html#L45

https://github.com/mkdocs/mkdocs/blob/273e3baa8e47fa5d121402d98757d3f5a5510383/mkdocs/themes/readthedocs/base.html#L47

https://github.com/mkdocs/mkdocs/blob/adfd76d66fc8b53197d8875d64e463d05a71d5a5/mkdocs/themes/mkdocs/base.html#L24

https://github.com/mkdocs/mkdocs/blob/adfd76d66fc8b53197d8875d64e463d05a71d5a5/mkdocs/themes/mkdocs/base.html#L36

https://github.com/mkdocs/mkdocs/blob/adfd76d66fc8b53197d8875d64e463d05a71d5a5/mkdocs/themes/mkdocs/base.html#L38

**Comment by fu-sen** [comment]
This was a question that users familiar with MkDocs could answer, even though it's been a few days.

```
  {%- if config.theme.highlightjs %}
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  {%- endif %}
```
```
  {%- if config.theme.highlightjs %}
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  {%- for lang in config.theme.hljs_languages %}
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/{{lang}}.min.js"></script>
  {%- endfor %}
```

Strictly speaking, it is highlight.js provided by [cdnjs](https://cdnjs.com/), It's Cloudflare provides a CDN.
but this can be disabled in the settings:

**mkdocs.yml**:
```
theme:
    name: mkdocs
    highlightjs: false
```

[Overriding Template Blocks](https://www.mkdocs.org/user-guide/styling-your-docs/#overriding-template-blocks) is required to delete Google Fonts as well.

**mkdocs.yml**:
```
theme:
  name: 'lumen'
  custom_dir: 'theme/'
```

**theme/main.html**:
```
{% extends "base.html" %}

  {%- block styles %}
  <link rel="stylesheet" href="{{ 'css/theme.css'|url }}" />
  <link rel="stylesheet" href="{{ 'css/theme_extra.css'|url }}" />
  {%- for path in config['extra_css'] %}
  <link href="{{ path|url }}" rel="stylesheet" />
  {%- endfor %}
  {%- endblock %}
```

If you want, you can use `extra_css` or` extra_js` to specify alternatives.

**Comment by oprypin** [comment]
The tricky part here is this `https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/{{lang}}.min.js`.
So basically all of the possible `lang` files would need to be vendored *and* included into every site output, because MkDocs has no mechanism to optionally include static files, as far as I'm aware.

**Comment by oprypin** [comment]
Maybe people have ideas for a way to solve that *well*. For now, perhaps we could handle only the easy case.

And another unfortunate thing I forgot to mention: even the base highlightjs file is optional as per the theme configuration, but it can't be made optional in the output.

**Comment by squidfunk** [comment]
If you're a user of [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/), enabling the [built-in privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#built-in-privacy-plugin) will automatically download and replace all [external assets](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#external-assets) that are found in the resulting HTML with local versions. This means that https://github.com/mkdocs/mkdocs/issues/2171#issuecomment-871565889 is also solved, since all external files found in the HTML are handled - no additional vendoring necessary.

**Comment by tomchristie** [comment]
Yep I think it'd be a good idea for the `readthedocs` and `mkdocs` themes to include all assets as part of the package.

**Comment by kkarhan** [comment]
I think the [privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#built-in-privacy-plugin) [as mentioned](https://github.com/mkdocs/mkdocs/issues/2171#issuecomment-1312447974) should be the default setting for most setups.

- Not only due to the privacy concerns ([CLOUD Act](https://en.wikipedia.org/wiki/CLOUD_Act) is *inherently and unfixably incompatible* with [GDPR](https://en.wikipedia.org/wiki/General_Data_Protection_Regulation) and [BDSG](https://en.wikipedia.org/wiki/Bundesdatenschutzgesetz) and I've yet to see any lawyer question that *at all*!) but also because not everyone wants to have their documentation on a network that has internet access. (In fact, I know of several companies in *crititcal infrastructure* settings that are per law obligated to have their systems [airgapped](https://en.wikipedia.org/wiki/Air_gap_(networking)) as part of mandatory security standards.

- And don't even get me started on the [entire shitshow](https://en.wikipedia.org/wiki/Cloudflare#Controversies) that is ClownFlare. They are [by all evidence](https://www.youtube.com/watch?v=Np9erdnM4l8&list=PLI84-gNHkUdvE1u_Kfz3qS9h75n-m-T3y) a [rogue ISP](https://en.wikipedia.org/wiki/Bulletproof_hosting) and should be treated as such!

**Comment by squidfunk** [comment]
Note that the privacy plugin is provided by Material for MkDocs, not MkDocs. Unfortunately, MkDocs does not allow themes to enable certain plugins by default, so it needs to be explicitly enabled. This might change in the future. It's also good to know, that the privacy plugin will work with any theme, not only with Material for MkDocs.
